models/trie_mimc/frontend_proof.go

```diff
@@ -0,0 +1,477 @@
+package trie_mimc
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/hash/mimc"
+)
+
+// Return the result of right shift by 1, input size = 32
+// Input: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231, 132}
+// Output: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231}
+func rightShift1(api frontend.API, input frontend.Variable) frontend.Variable {
+	var lsb8 frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	for i := 0; i < 8; i++ {
+		lsb8 = api.Add(lsb8, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	return api.DivUnchecked(api.Sub(input, lsb8), 256)
+}
+
+func leastNBytes(api frontend.API, input frontend.Variable, N int) frontend.Variable {
+	var lsb frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	if N >= 32 {
+		return api.Add(input, 0)
+	}
+	for i := 0; i < 8*N; i++ {
```

Comment:
similarly to the previous function. if divUnchecked removes the remainder then 
`func leastNBytes` could be reduced to 
`return api.Sub ( input , api.Mult ( api.DivUnchecked ( input, MaxMultiplier ) , MaxMultiplier ) )`

Reply:
Good point. Unfortunately DivUnchecked cannot remove the remainder so I will need to implement it like this.